labscommunity/protocol-land
# fix(contract): Update userRepoIdMap on forking

## Summary

This PR fixes issue with contract where a user can fork a repository with a name then can create a repository with same name later.

## How to Test
1) If you need to see if this issue exist, deploy a new contract from deployment branch (as it's already fixed in this PR & also don't test in the production contract). Fork a repository then create a repository with same name.
2) Then again Fork a repository then create a repository with same name with this PR to see if it's fixed.
3) Evolve the contract and also update `userRepoIdMap` state to update the missing repos data.

Here's a code to update `userRepoIdMap`

```ts

export async function updateUserRepoIdMapData(
  state: ContractState,
  { input: { payload }, caller }: RepositoryAction
): Promise<ContractResult<ContractState>> {
  const userRepoIdMap = state.userRepoIdMap
  for (const repoId in state.repos) {
    const repo = state.repos[repoId]

    const lowercasedRepoName = repo.name.toLowerCase()
    const ownerRepos = userRepoIdMap[repo.owner] ?? {}

    if (!ownerRepos[lowercasedRepoName]) {
      ownerRepos[lowercasedRepoName] = repo.id
    }

    userRepoIdMap[repo.owner] = ownerRepos
  }

  state.userRepoIdMap = userRepoIdMap

  return { state }
}
```
